### ISSUE TITLE
Logging Operator CRDs always out-of-sync

### ISSUE THREAD

**Issue by EppO** [argoproj/argo-cd#3391]
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a
question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [x] I've searched in the docs and FAQ for my answer: http://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

[Logging Operator](https://github.com/banzaicloud/logging-operator/) creates [CRDs](https://github.com/banzaicloud/logging-operator/tree/master/charts/logging-operator/crds) that Argo CD considers always out of sync. 

**To Reproduce**

Create Argo CD app that uses [Logging Operator Chart](https://github.com/banzaicloud/logging-operator/tree/master/charts/logging-operator)

**Expected behavior**

Argo CD app synced

**Screenshots**

```
argocd app get myapp | grep OutOfSync
Sync Status:        OutOfSync from SRED-206-Banzai-Operator (873959d)
apiextensions.k8s.io       CustomResourceDefinition               clusterflows.logging.banzaicloud.io                              OutOfSync
apiextensions.k8s.io       CustomResourceDefinition               clusteroutputs.logging.banzaicloud.io                            OutOfSync
apiextensions.k8s.io       CustomResourceDefinition               flows.logging.banzaicloud.io                                     OutOfSync
apiextensions.k8s.io       CustomResourceDefinition               loggings.logging.banzaicloud.io                                  OutOfSync
apiextensions.k8s.io       CustomResourceDefinition               outputs.logging.banzaicloud.io                                   OutOfSync
```

**Version**

```shell
argocd: v1.5.1+8a3b36b
  BuildDate: 2020-04-06T15:53:49Z
  GitCommit: 8a3b36bd28fe278f7407b9a8797c79ad859d1acd
  GitTreeState: clean
  GoVersion: go1.14
  Compiler: gc
  Platform: linux/amd64
argocd-server: v1.5.1+8a3b36b
  BuildDate: 2020-04-06T15:58:54Z
  GitCommit: 8a3b36bd28fe278f7407b9a8797c79ad859d1acd
  GitTreeState: clean
  GoVersion: go1.14
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: Version: {Version:kustomize/v3.2.1 GitCommit:d89b448c745937f0cf1936162f26a5aac688f840 BuildDate:2019-09-27T00:10:52Z GoOs:linux GoArch:amd64}
  Helm Version: version.BuildInfo{Version:"v3.1.1", GitCommit:"afe70585407b420d0097d07b21c47dc511525ac8", GitTreeState:"clean", GoVersion:"go1.13.8"}
  Kubectl Version: v1.14.0
```

**Logs**

```
# argocd app diff myapp
===== apiextensions.k8s.io/CustomResourceDefinition /clusteroutputs.logging.banzaicloud.io ======
5185,5199c5185,5188
<     kind: ClusterOutput
<     plural: clusteroutputs
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:49Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
===== apiextensions.k8s.io/CustomResourceDefinition /flows.logging.banzaicloud.io ======
647,668c647,650
<     kind: Flow
<     plural: flows
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: '[spec.validation.openAPIV3Schema.properties[spec].properties[filters].items.properties[parser].properties[parse].properties[patterns].items.type:
<       Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[filters].items.properties[parser].properties[parsers].items.properties[patterns].items.type:
<       Required value: must not be empty for specified array items]'
<     reason: Violations
<     status: "True"
<     type: NonStructuralSchema
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:49Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
===== apiextensions.k8s.io/CustomResourceDefinition /outputs.logging.banzaicloud.io ======
5179,5193c5179,5182
<     kind: Output
<     plural: outputs
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:43Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:49Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
===== apiextensions.k8s.io/CustomResourceDefinition /clusterflows.logging.banzaicloud.io ======
654,675c654,657
<     kind: ClusterFlow
<     plural: clusterflows
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:49Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: '[spec.validation.openAPIV3Schema.properties[spec].properties[filters].items.properties[parser].properties[parse].properties[patterns].items.type:
<       Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[filters].items.properties[parser].properties[parsers].items.properties[patterns].items.type:
<       Required value: must not be empty for specified array items]'
<     reason: Violations
<     status: "True"
<     type: NonStructuralSchema
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
===== apiextensions.k8s.io/CustomResourceDefinition /loggings.logging.banzaicloud.io ======
2673,2687c2673,2676
<     kind: Logging
<     plural: loggings
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:41Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:46Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
[root@svvacmn0001 ~]# argocd version
argocd: v1.5.1+8a3b36b
  BuildDate: 2020-04-06T15:53:49Z
  GitCommit: 8a3b36bd28fe278f7407b9a8797c79ad859d1acd
  GitTreeState: clean
  GoVersion: go1.14
  Compiler: gc
  Platform: linux/amd64
argocd-server: v1.5.1+8a3b36b
  BuildDate: 2020-04-06T15:58:54Z
  GitCommit: 8a3b36bd28fe278f7407b9a8797c79ad859d1acd
  GitTreeState: clean
  GoVersion: go1.14
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: Version: {Version:kustomize/v3.2.1 GitCommit:d89b448c745937f0cf1936162f26a5aac688f840 BuildDate:2019-09-27T00:10:52Z GoOs:linux GoArch:amd64}
  Helm Version: version.BuildInfo{Version:"v3.1.1", GitCommit:"afe70585407b420d0097d07b21c47dc511525ac8", GitTreeState:"clean", GoVersion:"go1.13.8"}
  Kubectl Version: v1.14.0
[root@svvacmn0001 ~]# argocd app diff infra-stack
===== apiextensions.k8s.io/CustomResourceDefinition /clusterflows.logging.banzaicloud.io ======
654,675c654,657
<     kind: ClusterFlow
<     plural: clusterflows
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:49Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: '[spec.validation.openAPIV3Schema.properties[spec].properties[filters].items.properties[parser].properties[parse].properties[patterns].items.type:
<       Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[filters].items.properties[parser].properties[parsers].items.properties[patterns].items.type:
<       Required value: must not be empty for specified array items]'
<     reason: Violations
<     status: "True"
<     type: NonStructuralSchema
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
===== apiextensions.k8s.io/CustomResourceDefinition /outputs.logging.banzaicloud.io ======
5179,5193c5179,5182
<     kind: Output
<     plural: outputs
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:43Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:49Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
===== apiextensions.k8s.io/CustomResourceDefinition /flows.logging.banzaicloud.io ======
647,668c647,650
<     kind: Flow
<     plural: flows
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: '[spec.validation.openAPIV3Schema.properties[spec].properties[filters].items.properties[parser].properties[parse].properties[patterns].items.type:
<       Required value: must not be empty for specified array items, spec.validation.openAPIV3Schema.properties[spec].properties[filters].items.properties[parser].properties[parsers].items.properties[patterns].items.type:
<       Required value: must not be empty for specified array items]'
<     reason: Violations
<     status: "True"
<     type: NonStructuralSchema
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:49Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
===== apiextensions.k8s.io/CustomResourceDefinition /loggings.logging.banzaicloud.io ======
2673,2687c2673,2676
<     kind: Logging
<     plural: loggings
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:41Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:46Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
===== apiextensions.k8s.io/CustomResourceDefinition /clusteroutputs.logging.banzaicloud.io ======
5185,5199c5185,5188
<     kind: ClusterOutput
<     plural: clusteroutputs
<   conditions:
<   - lastTransitionTime: "2020-04-09T16:39:44Z"
<     message: no conflicts found
<     reason: NoConflicts
<     status: "True"
<     type: NamesAccepted
<   - lastTransitionTime: "2020-04-09T16:39:49Z"
<     message: the initial names have been accepted
<     reason: InitialNamesAccepted
<     status: "True"
<     type: Established
<   storedVersions:
<   - v1beta1
---
>     kind: ""
>     plural: ""
>   conditions: []
>   storedVersions: []
```

**Comment by alexmt** [comment]
Hello @EppO ,

Looks like CRDs in logging-operator has status in the manifests. E.g. https://github.com/banzaicloud/logging-operator/blob/master/charts/logging-operator/crds/logging.banzaicloud.io_clusterflows.yaml#L777

The CRD status is being changed by the Kubernetes and Argo CD detects it as a deviation. I'm not sure why but looks like a lot of CRDs include empty status. Probably to reset CRD condition every time when CRD get changed. The workaround is to use [diffing customization](https://argoproj.github.io/argo-cd/user-guide/diffing/) to ignore CRD status:

```yaml
  ignoreDifferences:
    - group: apiextensions.k8s.io
      kind: CustomResourceDefinition
      jsonPointers:
        - /status
```

**Comment by alexmt** [comment]
Looks like this is a common issue for multiple helm charts. Created the ticket to ignore CRD status by default: https://github.com/argoproj/argo-cd/issues/2239

**Comment by EppO** [comment]
addressed here #3393
